File: src/bot/commands/war/lineup.ts
import { AutocompleteInteraction, SlashCommandBuilder, type ChatInputCommandInteraction } from 'discord.js';
import { Command } from '../../struct/Command.js';
import { baseEmbed } from '../../../util/embeds.js';
import { pad, padStart } from '../../../util/helper.js';

/** War lineup — both sides ordered by map position with Town Hall levels. */
export default class LineupCommand extends Command {
  public constructor() {
    super('lineup', { category: 'war', description: 'Show the war lineup for both sides.', defer: true });
  }

  public builder() {
    return new SlashCommandBuilder()
      .setName('lineup')
      .setDescription('Show the war lineup for both sides.')
      .addStringOption((opt) => opt.setName('tag').setDescription('Clan tag or alias.').setAutocomplete(true));
  }

  public async autocomplete(interaction: AutocompleteInteraction, args: Record<string, any>) {
    await interaction.respond(await this.client.resolver.clanAutocomplete(interaction.guildId, String(args.tag ?? '')));
  }

  private side(members: any[]): string {
    return members
      .sort((a, b) => a.mapPosition - b.mapPosition)
      .map((m, i) => `\`${padStart(i + 1, 2)} TH${padStart(m.townHallLevel, 2)} ${pad(m.name, 15)}\``)
      .join('\n')
      .slice(0, 1024);
  }

  public async exec(interaction: ChatInputCommandInteraction, args: Record<string, any>) {
    const { data: clan, error } = await this.client.resolver.resolveClan(interaction, args.tag);
    if (!clan) return interaction.editReply(error!);

    const war = await this.client.coc.getCurrentWar(clan.tag);
    if (!war || war.state === 'notInWar') return interaction.editReply('Clan is not currently in a war.');
    const w = war as any;

    const embed = baseEmbed(this.client, interaction)
      .setTitle(`${w.clan.name} vs ${w.opponent.name} — Lineup`)
      .addFields(
        { name: w.clan.name, value: this.side(w.clan.members), inline: true },
        { name: w.opponent.name, value: this.side(w.opponent.members), inline: true }
      );

    return interaction.editReply({ embeds: [embed] });
  }
}

File: cogs/war/commands/lineup.py
"""/war lineup, both sides ordered by map position."""

import coc
import discord

from utils.embeds import base_embed, error_embed
from utils.helpers import pad, pad_start
from utils.resolver import resolve_clan


def _side(members) -> str:
    ordered = sorted(members, key=lambda m: m.map_position)
    lines = [f"`{pad_start(i, 2)} TH{pad_start(m.town_hall, 2)} {pad(m.name, 15)}`" for i, m in enumerate(ordered, 1)]
    return "\n".join(lines)[:1024]


async def lineup(interaction: discord.Interaction, tag: str = None):
    await interaction.response.defer()
    try:
        clan = await resolve_clan(interaction, tag)
    except ValueError as err:
        await interaction.followup.send(embed=error_embed(str(err)))
        return

    try:
        war = await interaction.client.coc_client.get_current_war(clan.tag)
    except coc.PrivateWarLog:
        await interaction.followup.send(embed=error_embed("This clan's war log is private."))
        return
    except Exception:
        await interaction.followup.send(embed=error_embed("Clash of Clans API request failed, try again in a bit."))
        return

    if war is None or war.state == "notInWar":
        await interaction.followup.send(embed=error_embed(f"{clan.name} is not currently in a war."))
        return

    embed = await base_embed(interaction, title=f"{war.clan.name} vs {war.opponent.name}, Lineup")
    embed.add_field(name=war.clan.name, value=_side(war.clan.members), inline=True)
    embed.add_field(name=war.opponent.name, value=_side(war.opponent.members), inline=True)
    await interaction.followup.send(embed=embed)
